### PR TITLE
coprocessor: decode string with given charset in field type

### PR DESCRIPTION
**Pull request by hicqu**
## What have you changed? (mandatory)
Some place we decode string by calling `String::from_utf8`, but it's not correctly. Strings should be decoded with given charset from TiDB in `field_type`.

This PR fix the potential bug. And the dependencency `Regex` is upgrade to `0.2` because `0.1` can't support unicode pattern in `regex::bytes::Regex`.

## What are the type of the changes? (mandatory)
Fix and improvement.

## How has this PR been tested? (mandatory)
make dev.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No.

## Does this PR affect tidb-ansible update? (mandatory)
No.

## Add a few positive/negative examples (optional)
Let's talk about `regexp`. The behavior is different from `MySQL`, but same with `TiDB`:

In MySQL:
```sql
mysql> select cast('你' as binary) regexp '^.$';
+------------------------------------+
| cast('你' as binary) regexp '^.$'  |
+------------------------------------+
|                                  0 |
+------------------------------------+
1 row in set (0.00 sec)
```

In TiDB:
```sql
mysql> select cast('你' as binary) regexp '^.$';
+------------------------------------+
| cast('你' as binary) regexp '^.$'  |
+------------------------------------+
|                                  1 |
+------------------------------------+
```

I think MySQL behavior is more reasonable. Because if the column's field type is binary, we can't treat it as UTF8 by default. But because `TiKV` and `TiDB` use third party libraries to do this, which are not designed for this case. Maybe later we need to consider to keep compatible with MySQL.
/cc @zz-jason

### CODE REVIEW

**Comment by breeswish** [comment]
Seems that we are currently use &str everywhere, which forces the use of utf-8, so that even in future a non-utf8 charset is given, it cannot be supported?

**Comment by breeswish** [comment]
@hicqu Got it.

File: src/coprocessor/dag/expr/builtin_like.rs [hunk 1]
```diff
@@ -33,26 +32,20 @@ impl ScalarFunc {
...
         let target = try_opt!(self.children[0].eval_string(ctx, row));
-        let pattern = try_opt!(self.children[1].eval_string(ctx, row));
-
-        let p = str::from_utf8(&pattern)?;
-        let t = str::from_utf8(&target)?;
+        let pattern = try_opt!(self.children[1].eval_string_and_decode(ctx, row));
 
         // TODO: cache compiled result
-        Ok(Some(Regex::new(p)?.is_match(t) as i64))
+        Ok(Some(BytesRegex::new(&pattern)?.is_match(&target) as i64))
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_like.rs]: Seems `target` should be evaluated by `eval_string_and_decode` too?

Comment (hicqu) [reply src/coprocessor/dag/expr/builtin_like.rs]: In `regexp` yes, but `regexp_binary` not because the `target` could be an invalid `utf8` bytes.

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**